File: Treachery-iOS/Treachery-iOSTests/CardDatabaseTests.swift
import Testing
@testable import Treachery_iOS

struct CardDatabaseTests {

    // MARK: - Loading

    @Test func loadsCards() {
        let db = CardDatabase.shared
        #expect(!db.allCards.isEmpty)
    }

    @Test func has62Cards() {
        #expect(CardDatabase.shared.allCards.count == 62)
    }

    // MARK: - Role Filtering

    @Test func correctLeaderCount() {
        let leaders = CardDatabase.shared.cards(forRole: .leader)
        #expect(leaders.count == 13)
        #expect(leaders.allSatisfy { $0.role == .leader })
    }

    @Test func correctGuardianCount() {
        let guardians = CardDatabase.shared.cards(forRole: .guardian)
        #expect(guardians.count == 18)
        #expect(guardians.allSatisfy { $0.role == .guardian })
    }

    @Test func correctAssassinCount() {
        let assassins = CardDatabase.shared.cards(forRole: .assassin)
        #expect(assassins.count == 18)
        #expect(assassins.allSatisfy { $0.role == .assassin })
    }

    @Test func correctTraitorCount() {
        let traitors = CardDatabase.shared.cards(forRole: .traitor)
        #expect(traitors.count == 13)
        #expect(traitors.allSatisfy { $0.role == .traitor })
    }

    @Test func roleCountsSumToTotal() {
        let db = CardDatabase.shared
        let total = Role.allCases.reduce(0) { $0 + db.cards(forRole: $1).count }
        #expect(total == db.allCards.count)
    }

    // MARK: - Lookup

    @Test func findCardById() {
        let db = CardDatabase.shared
        guard let first = db.allCards.first else {
            Issue.record("No cards loaded")
            return
        }
        let found = db.card(withId: first.id)
        #expect(found?.id == first.id)
        #expect(found?.name == first.name)
    }

    @Test func returnsNilForMissingId() {
        #expect(CardDatabase.shared.card(withId: "nonexistent") == nil)
    }

    // MARK: - Rarity Filtering

    @Test func rarityFilterReturnsCorrectCards() {
        let db = CardDatabase.shared
        for rarity in Rarity.allCases {
            let cards = db.cards(forRarity: rarity)
            #expect(cards.allSatisfy { $0.rarity == rarity })
        }
    }

    @Test func rarityCountsSumToTotal() {
        let db = CardDatabase.shared
        let total = Rarity.allCases.reduce(0) { $0 + db.cards(forRarity: $1).count }
        #expect(total == db.allCards.count)
    }

    // MARK: - Random Selection

    @Test func randomCardsRespectsCount() {
        let cards = CardDatabase.shared.randomCards(forRole: .assassin, count: 3)
        #expect(cards.count == 3)
        #expect(cards.allSatisfy { $0.role == .assassin })
    }

    @Test func randomCardsReturnsUniqueCards() {
        let cards = CardDatabase.shared.randomCards(forRole: .leader, count: 5)
        let ids = Set(cards.map(\.id))
        #expect(ids.count == cards.count)
    }

    @Test func randomCardsClampedToAvailable() {
        let cards = CardDatabase.shared.randomCards(forRole: .leader, count: 100)
        #expect(cards.count == 13) // Only 13 leaders exist
    }

    // MARK: - Card Data Integrity

    @Test func allCardsHaveUniqueIds() {
        let ids = CardDatabase.shared.allCards.map(\.id)
        #expect(Set(ids).count == ids.count)
    }

    @Test func allCardsHaveUniqueCardNumbers() {
        let numbers = CardDatabase.shared.allCards.map(\.cardNumber)
        #expect(Set(numbers).count == numbers.count)
    }

    @Test func allCardsHaveNonEmptyNames() {
        for card in CardDatabase.shared.allCards {
            #expect(!card.name.isEmpty)
        }
    }

    @Test func allCardsHaveNonEmptyAbilityText() {
        for card in CardDatabase.shared.allCards {
            #expect(!card.abilityText.isEmpty)
        }
    }
}

File: TreacheryAndroid/app/src/test/java/com/solomon/treachery/data/CardDatabaseTests.kt
package com.solomon.treachery.data

import com.solomon.treachery.model.IdentityCard
import com.solomon.treachery.model.Rarity
import com.solomon.treachery.model.Role
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

/**
 * Tests for the card database using the bundled identity_cards.json.
 * Mirrors iOS CardDatabaseTests.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class CardDatabaseTests {

    private val json = Json { ignoreUnknownKeys = true }
    private lateinit var allCards: List<IdentityCard>

    @BeforeAll
    fun loadCards() {
        val jsonString = this::class.java.classLoader!!
            .getResourceAsStream("identity_cards.json")!!
            .bufferedReader()
            .use { it.readText() }
        allCards = json.decodeFromString<List<IdentityCard>>(jsonString)
    }

    // ── Loading ──

    @Nested
    inner class LoadingTests {
        @Test
        fun `loads 62 cards`() {
            assertEquals(62, allCards.size)
        }
    }

    // ── Role Filtering ──

    @Nested
    inner class RoleFilteringTests {
        @Test
        fun `13 leaders`() {
            assertEquals(13, allCards.count { it.role == Role.LEADER.value })
        }

        @Test
        fun `18 guardians`() {
            assertEquals(18, allCards.count { it.role == Role.GUARDIAN.value })
        }

        @Test
        fun `18 assassins`() {
            assertEquals(18, allCards.count { it.role == Role.ASSASSIN.value })
        }

        @Test
        fun `13 traitors`() {
            assertEquals(13, allCards.count { it.role == Role.TRAITOR.value })
        }

        @Test
        fun `all cards have a valid role`() {
            allCards.forEach { card ->
                assertNotNull(Role.fromValue(card.role), "Card '${card.name}' has invalid role '${card.role}'")
            }
        }
    }

    // ── Lookup ──

    @Nested
    inner class LookupTests {
        @Test
        fun `find card by id`() {
            val first = allCards.first()
            val found = allCards.find { it.id == first.id }
            assertNotNull(found)
            assertEquals(first.name, found!!.name)
        }

        @Test
        fun `unknown id returns null`() {
            val found = allCards.find { it.id == "nonexistent-id" }
            assertNull(found)
        }
    }

    // ── Rarity Filtering ──

    @Nested
    inner class RarityFilteringTests {
        @Test
        fun `all cards have a valid rarity`() {
            allCards.forEach { card ->
                assertNotNull(Rarity.fromValue(card.rarity), "Card '${card.name}' has invalid rarity '${card.rarity}'")
            }
        }

        @Test
        fun `each rarity has at least one card`() {
            Rarity.entries.forEach { rarity ->
                assertTrue(
                    allCards.any { it.rarity == rarity.value },
                    "No cards found for rarity ${rarity.value}"
                )
            }
        }
    }

    // ── Random Selection ──

    @Nested
    inner class RandomSelectionTests {
        @Test
        fun `random selection respects count`() {
            val leaderCards = allCards.filter { it.role == Role.LEADER.value }
            val selected = leaderCards.shuffled().take(3)
            assertEquals(3, selected.size)
        }

        @Test
        fun `random selection never exceeds available`() {
            val leaderCards = allCards.filter { it.role == Role.LEADER.value }
            val selected = leaderCards.shuffled().take(100)
            assertEquals(leaderCards.size, selected.size)
        }

        @Test
        fun `random selection returns unique cards`() {
            val selected = allCards.filter { it.role == Role.ASSASSIN.value }.shuffled().take(5)
            assertEquals(selected.size, selected.distinctBy { it.id }.size)
        }
    }

    // ── Data Integrity ──

    @Nested
    inner class DataIntegrityTests {
        @Test
        fun `all cards have unique IDs`() {
            val ids = allCards.map { it.id }
            assertEquals(ids.size, ids.toSet().size, "Duplicate card IDs found")
        }

        @Test
        fun `all cards have unique card numbers`() {
            val numbers = allCards.map { it.cardNumber }
            assertEquals(numbers.size, numbers.toSet().size, "Duplicate card numbers found")
        }

        @Test
        fun `all cards have non-empty names`() {
            allCards.forEach { card ->
                assertTrue(card.name.isNotBlank(), "Card ${card.id} has blank name")
            }
        }

        @Test
        fun `all cards have non-empty ability text`() {
            allCards.forEach { card ->
                assertTrue(card.abilityText.isNotBlank(), "Card '${card.name}' has blank ability text")
            }
        }

        @Test
        fun `all cards have non-empty unveil cost`() {
            allCards.forEach { card ->
                assertTrue(card.unveilCost.isNotBlank(), "Card '${card.name}' has blank unveil cost")
            }
        }
    }
}
